fix msbuild commandline when executing msbuild from within autotools toolchain

chr-thien opened the pull request:
Changelog: Omit
Docs: Omit

Fixes: #15627 

Fixes msbuild commandline when executing msbuild from within autotools toolchain.

BOPT-ct committed:
Update msbuild.py

fix msbuild commandline when executing within autotools toolchain

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=18005) <br/>All committers have signed the CLA.

memsharded reviewed (commented):
Hi @chr-thien 

Thanks for your contribution.
It would be necessary to have first a failing test in the test suite, to evidence what is breaking and guarantee the fix works and it doesn't break in the future.

Please check existing tests and don't hesitate to ask for help or guidance if you need it.

BOPT-ct committed:
adapt msbuild unittests

BOPT-ct committed:
adapt msbuild integration tests

chr-thien commented:
OK.
We need a minimal recipe with this:
`def generate(self):`
`        tc = AutotoolsToolchain(self)`
`def build(self):`
`        msbuild = MSBuild(self)`
`        msbuild.build(sln)`

The test could be a new test in directory test\functional\toolchains\gnu\autotools\ or it could be appended to test_win_bash.py in very same directory.

I'll still need to investigate on test data files (e.g. do we have a minimal MSBuild solution+project available?).

memsharded commented:
> The test could be a new test in directory test\functional\toolchains\gnu\autotools\ or it could be appended to test_win_bash.py in very same directory.

As you prefer, there are also some related tests in ``test_basic.py`` in that directory.

> def generate(self):
        tc = AutotoolsToolchain(self)
def build(self):
        msbuild = MSBuild(self)
        msbuild.build(sln)

This setup is not very clear, how is it possible to have an ``AutotoolsToolchain`` and use ``MSBuild``? 
This is not expected, ``MSBuild`` is intended to work with ``MSBuildToolchain`` and ``MSBuildDeps``, and ``AutotoolsToolchain`` is expected to work with ``Autotools`` wrapper. Could you clarify this?

> I'll still need to investigate on test data files (e.g. do we have a minimal MSBuild solution+project available?).

Yes, there are tests in ``functional/toolchains/microsoft``, and there is even the ``conan new msbuild_lib/msbuild_exe`` templates.

chr-thien commented:
> This setup is not very clear, how is it possible to have an `AutotoolsToolchain` and use `MSBuild`? This is not expected, `MSBuild` is intended to work with `MSBuildToolchain` and `MSBuildDeps`, and `AutotoolsToolchain` is expected to work with `Autotools` wrapper. Could you clarify this?

Imagine a library using `make` for building on all platforms and output is executable or shared library.

After building with `make` one wants to create an installer module. Especially on Windows one could use Windows Installer XML - which is a MSBuild project. Voilà! Invoke `msbuild.exe` from Autotools environment. The installer module then becomes part of the package. The installer module to an installer is like a shared library to an application.

See also #15627 and comment https://github.com/conan-io/conan/issues/15627#issuecomment-1935679122

memsharded commented:
> Invoke msbuild.exe from Autotools environment.

Isn't this accidental?
I mean, wouldn't the ``msbuild`` execution run the same from within a Windows cmd terminal or regular process, not from Autotools environment? I am not familiar with that ``Windows Installer XML``, but I'd say it is likely that the env-vars defined for the ``make`` to run aren't used by the ``msbuild`` project.

If that is the case, it might be possible to do something like:

```python
def build(self):
    autotools.configure()
    autotools.make()
    self.win_bash = False
    msbuild = MSBuild(self)
    msbuild.build("project.sln")
```

I am trying to fully understand the reasons and all the possibilities, because I am concerned about these kind of changes. I know it doesn't look like very high risk, but my experience and scars from many years maintaining this project tell me that it is very likely that these changes will break other users in some setups that will not like the ``-`` instead of the MSBuild canonical ``/``, even if ``-`` is accepted.

chr-thien committed:
adapt msbuild functional test

chr-thien commented:
> ```python
> self.win_bash = False
> ```

After adding that I get this error:

ERROR: icu/73.2@opus/development: Error in build() method, line 244
        msbuild.build(os.path.join(self.build_folder, "..", "msbuild_wix", "icu.sln"))
        ConanException: Cannot wrap command with different envs,['C:\\Users\\christoph.thien\\.conan2\\p\\b\\icuc566cdc15937b\\b\\build-debug\\conan\\conanbuild.bat'] - ['C:\\Users\\christoph.thien\\.conan2\\p\\b\\icuc566cdc15937b\\b\\build-debug\\conan\\conanbuild.sh']



> ```
> I mean, wouldn't the msbuild execution run the same from within a Windows cmd terminal or regular process, not from Autotools environment? I am not familiar with that Windows Installer XML, but I'd say it is likely that the env-vars defined for the make to run aren't used by the msbuild project.
> ```

I agree. The environment prepared for `make` is not needed for `msbuild`.

So I see two approaches here to make it work:
Either let `msbuild` successfully execute under Autotools toolchain (this PR makes that possible).
Or run `msbuild` after `make` in its native toolchain - which has to be specified somehow or toolchain has to be switched.



I think the change from `/` to `-` for commandline arguments will not do harm.
Changing `msbuild` to `msbuild.exe` has bigger potential of introducing errors.

chr-thien committed:
add functional test

This functional test runs a recipe with autotools toolchain to build a simple application. Then runs msbuild under autotools toolchain to copy the application to another directory. The functional test shall verify correct invocation of msbuild from autotools environment.

chr-thien commented:
I added a functional test named 'test_msys2_and_msbuild'.
That test fails when changes from this PR are not present.
The test succeeds when using `msbuild.exe` and `-` for commandline arguments.

memsharded commented:
Thanks for the tests and following up.

I am still a bit concern about the possibility of this breaking, but I am adding it to 2.16 to try to move it forward (not guaranteed)

memsharded commented on conan/tools/microsoft/msbuild.py:
```diff
@@ -13,7 +13,7 @@ def msbuild_verbosity_cmd_line_arg(conanfile):
             "verbose": "Detailed",
         }.get(verbosity)
-        return f'/verbosity:{verbosity}'
+        return f'-verbosity:{verbosity}'
```
It seems that ``-`` is supported long time ago, apparently at least VS 2013, so this should be safe
